### DISCUSSION TITLE
Lesson 11 Dogie NFT blank

### DISCUSSION THREAD

**Discussion by DukeLurkin** [Cyfrin/foundry-full-course-cu#1057] [Q&A] [answered]
For some reason the Dogie Nft I created shows up as a blank image in Metamask.  The image is in VScode same a patricks, not sure what happened?
<img width="265" alt="blankDogieNft" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/47e95027-053e-44b2-a1b6-6d3bb1531d86">

**Comment by DukeLurkin** [comment]
Here is the link to my repo: https://github.com/DukeLurkin/foundry-nft-f23 
I've tried a bunch of different links for the PUG but nothing makes the image show.
<img width="254" alt="allBlankDogie" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/63a8aa21-fefc-4def-b3b8-5f4cf19ac41e">
<img width="683" alt="pugPNGLink" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/03e9a327-9d35-4ff6-8728-07fee4365d52">

**Reply by theirrationalone** [reply]
Please try these two mitigations....

1. try to remove string after `?` at the end of `PUG` uri, including `?`. and try again. or...
2. follow method 1 and let's trim the `PUG`'s string bit more... try to remove `https://ipfs.io/ipfs` from the beginning and replace it with string `ipfs://` and keep afterward appearing string upto `?` but don't include `?` as it is, into `PUG` string. Your final `PUG` should look like this: `ipfs://QmSsYRx3LpDAb1GZQm7zZ1AuHZjfbPkD6J7s9r41xu1mf8`. After all of this, Please try again.

method 2 has more potential to mitigate your issue.

;`)

**Reply by DukeLurkin** [reply]
I've tried all that with Patricks link in the repo and with my own ipfs links and nothing works.  Maybe I'm not clicking the right links???  So, after running "make deploy ARGS="--network sepolia"" I let it finish and click the link in this case: https://sepolia.etherscan.io/address/0xcf467f700a1ea4bce4d5ae2806b46da89ea132ad
<img width="926" alt="makeDeploylink" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/b27eb6a6-3c05-484e-9f10-49cd3c916825">

That link takes me here where I copy this contract address:
<img width="523" alt="makeDeploycontract" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/0ce5ce32-5c03-4797-afbd-7f023d801237">

I then click import NFT inside meta mask, paste that contract and use 0(zero) for token ID.  Is that correct??  I've also tried to follow the link after running "make mint ARGS=..." but that tells me I'm not the owner so maybe I did something else wrong?

**Reply by theirrationalone** [reply]
You're missing to put your `privateKey` aka `deployerKey` as a parameter inside `vm.startBroadcast(deployerKey)`.... As a result the calculated address of `interactions.s.sol` is the owner of the `tokenId` = 0.

**Reply by DukeLurkin** [reply]
Patrick does not have that in his code at this point.  I don't think that is the issue.  I don't even have a deployerKey variable created.  Thanks for trying though.  I just watched up to the point where he shows you the NFT in MM again, seems like im doing all the steps right.  I even tried a totally different picture, uploaded it to my ipfs and no matter how I cut up the link it's the same blank picture.

**Reply by DukeLurkin** [reply]
I tried to add a deployerKey anyway by following how Patrick has them in the repo and this is what I get back:
after running make deploy:
<img width="923" alt="deployerkeyadded1" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/7898e371-aae4-47f1-b80c-28ca7787c4bb">

after running make mint:

<img width="623" alt="deployerkeyadded2" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/713b8e5c-b0a6-436a-b2b0-4178b26bf71f">


I really hope im not stuck here for 2 months like I was with the raffle.  That would really really really fucking suck.

**Reply by DukeLurkin** [reply]
The private key is NOT 0 so I don't know what to do with that info.

**Reply by theirrationalone** [reply]
Your code is all okay... 

If you're facing issues with this then we can try `cast` in terminal.... 


Try this.... 


```bash
cast send 0xcf467f700A1eA4bce4d5ae2806b46Da89EA132aD "mintNft(string)" "https://ipfs.io/ipfs/QmSsYRx3LpDAb1GZQm7zZ1AuHZjfbPkD6J7s9r41xu1mf8" --rpc-url $SEPOLIA_RPC_URL --privateKey $PRIVATEKEY --broadcast
```

This will mint the nft with the tokenId = 1 if you haven't minted any other after that 0 tokenId, with your wallet as the tokenId holder... 



Make sure you have `SEPOLIA_RPC_URL` & `PRIVATEKEY` vars in your `.env` file... 

If don't please add them privateKey must be your metamask privateKey. 

And don't forget to run this command given below after adding data into `.env` file..... 

```solidity
source  .env
```

Test this method if it is working Or not... 


Then try this.... 

Update your pug variable:
PUG = `https://ipfs.io/ipfs/QmSsYRx3LpDAb1GZQm7zZ1AuHZjfbPkD6J7s9r41xu1mf8`

You Have to deploy your contract again if you modify it. 
 
```solidity
forge script  ./script/Interactions.s.sol:MintBasicNft --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATEKEY --broadcast
```

**Reply by theirrationalone** [reply]
There is a semantic bug in your makefile... 


If we're minting our nft then we don't need to verify our contract again... 

And NETWORK_ARGS has `--verify` command included. 

Try this specific commands chord for minting....


```solidity
forge script  ./script/Interactions.s.sol:MintBasicNft --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATEKEY --broadcast
```

**Reply by DukeLurkin** [reply]
I'll try all that, just wondering where did the address in cast send come from? this: "0xcf467f700A1eA4bce4d5ae2806b46Da89EA132aD "

**Reply by DukeLurkin** [reply]
Nvm! I see that address is one of my previous deployments now.  I've been running both "make deploy" and "make mint" every time to test the different ipfs location I set PUG to, should I only be rerunning "make mint"?

**Reply by theirrationalone** [reply]
Yes  off course you should only mint again.... Therefore you have been deployed your contract already... 

Aside minting, If you have made changes in your contract then as a side effect you would have to deploy your contract again.

**Reply by DukeLurkin** [reply]
So, I followed all that but running "forge script  ./script/Interactions.s.sol:MintBasicNft --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATEKEY --broadcast" didn't actually mint an NFT.  I reran mint using "make mint" and there was an NFT created at the corresponding address, it was even ID 1 instead of 0 but still blank:
<img width="239" alt="dogie1" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/c0ecb9be-ae50-42fe-a22b-3b1911d4d047">

This is what comes up after running your forge script:
<img width="1200" alt="forgeScriptIrrationalOne" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/5adadb75-0d45-432b-be23-e1a87eccf2a1">

I originally had just "--broadcast" but it gave me an error and said to put 2 --'s "-- --broadcast" in so I did, IDK lol

**Reply by DukeLurkin** [reply]
Okay, that's good to hear.  I figured if I changed the PUG ipfs address I'd need to reDeploy and that's what I was doing....I've redeployed many many times.

**Reply by theirrationalone** [reply]
Okh enough is enough 😅, i'm gonna give it a try . wait ....

**Reply by theirrationalone** [reply]
Heyyyy, @DukeLurkin !, Sorrrrryyyyyyy for providing only image ipfs link....

look i just minted & imported PUG NFT with  your NFT contract .....
![Screenshot (183)](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/d46331a7-c781-447c-a89c-9a2314304840)

And the mistake is minting wrong tokenURI....

The right one is this: `ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json`

Your whole contract is all Good, as an optional work around... You're not checking if that's a valid tokenURI or not...

here...
```solidity
function tokenURI(
        uint256 tokenId
      ) public view override returns (string memory) {
         if (!_exists(tokenId) {
            revert ContractName__tokenUriNotExists();
        }
          return s_tokenIdToUri[tokenId];
}
```

I'm lazzyyy Bad bad bad dev 😅, Sorry for inconvenience and confusing us both 😅😅😅

;`)

**Reply by DukeLurkin** [reply]
That's Patrick's tokenURI.  It's the first one I tried.  I tried uploading my own images to an ipfs, that's the PUG you see in my repo and it's still the same issue.  Maybe I'm missing a step??  I'll go one by one: So I plugged in that PUG ipfs you suggested from the foundry repo and ran make deploy:

<img width="834" alt="makeDeployIrrational" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/53630f66-6898-4f3c-a348-991299d2ec64">

Then I followed the link to the etherscan address that make deploy gave me:

<img width="923" alt="makeDeployIrrational2" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/41202c58-d1c5-415b-914f-8ff1ae8294bf">


Here:


<img width="841" alt="makeDeployIrrational3" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/3711a145-b272-4d0a-a2e2-839a1399ad8d">


I came back to VScode and ran make mint:


<img width="965" alt="makeDeployIrrational4" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/980c802e-f32c-4456-b8d2-76b8a09b00c4">

Once minting was done, I went back to the etherscan page that Deploy gave me and copied the address. 
I pasted the address into Import NFT in my meta mask along with the tokenID which was actually 1 not 0:

<img width="1254" alt="makeDeployIrrational6" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/8546bc65-e55b-4453-92d4-16651e8e3aa7">

It says NFT was added successfully but the picture is white.  You can see the description though, it says "adorable PUG pup" and has the same address and token ID all those details came along but the image :

<img width="1202" alt="makeDeployIrrational7" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/b705ea2a-ca56-43ee-9c81-c5d7ebf42ccb">


It seems like this happens no matter what I call the PUG tokenURI.

**Reply by theirrationalone** [reply, marked as answer]
You're doing it all right, and if you wanna pin your own URI into ipfs then you have to first pin your image to an uri pinning service i.e., pinata, [other service providers](https://www.jsonapi.co/media-storage/Pinata%20IPFS/alternatives), etc, which will make your IPFS URI up and running....

then you will need to make a json file containing data with properties as patrick's uri has. After that you will need to pin that json file to one of those services....

And It's really annoying... why it's not showing you image.... believe me you're doing everything right. You can now go for new lesson, It's all done already.

One more thing you can do is to clear your metamask's activity tab data and then try again. It'll be your luck if this triage works otherwise you're  101% all done with this.

**Reply by DukeLurkin** [reply]
OMFG!!!  I cleared my metamask activity:

<img width="261" alt="clearActivityTabData" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/4581ad7d-af52-4acd-ae3f-ebcd871d9465">

Then I enabled OpenSea API and AutoDetect NFTs:

<img width="256" alt="enableOpenSeaAndAutoDetectNFTs" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/92ff72fe-8fb7-4564-bb27-7e218796ff42">

And here it is! 

<img width="255" alt="dogieInMetamask" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/5d6385e5-37a1-440a-a25b-195b38009376">

Thank you so much again @theirrationalone !!!

**Reply by theirrationalone** [reply]
Great work @DukeLurkin ! 🚀🚀🚀

**Reply by DukeLurkin** [reply]
I got the ipfs source to work now too!  Thanks for the steps on that.  Took me a minute.  I didn't need to use pinata just so you know.  I uploaded it right to the ipfs.io Patrick has us install.  I copied his format json from PUG, put it in word, changed the variables and saved it as a json.  Then I uploaded that file to my ipfs node

<img width="1589" alt="ipfs1" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/0af3082d-749f-4e3e-b5b3-fcb453796b80">


, clicked "publish to IPNS"(not sure that was necessary) copied the "share link" BUT had to remove the ? and everything after it, 

<img width="1208" alt="ipfs2" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/df31cb9c-a800-43e9-a957-fac492cbb64c">


then pasted that for PUG tokenURI in VScode 

<img width="623" alt="ipfs3" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/a2b38c48-e7ba-4503-acb0-fdafa5162008">


and it worked!

<img width="1316" alt="ipfs4" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/123668611/d2e06a9d-30e2-45c0-ba44-1f746093b814">

**Reply by theirrationalone** [reply]
Wowww Great work @DukeLurkin ! 🚀🚀🚀

**Reply by Irwino820** [reply]
@DukeLurkin Hey, Can you please upload your repository?

I have weeks in this, no matter what i do, what i change, the picture doesnt apper :ccc

**Reply by DukeLurkin** [reply]
Hey @Irwino820, sorry this reply took so long.  here is the link to my repo: https://github.com/DukeLurkin/foundry-nft-f23
 but it was actually a problem with MetaMask settings for me.  If the repo is no help try these steps right after I say OMFG:  https://github.com/Cyfrin/foundry-full-course-f23/discussions/1057#discussioncomment-7448880

**Reply by Ynyesto** [reply]
I have the same issue. I tried enabling OpenSea and automatic detection but nothing seems to work... Any ideas what may be causing this? I believe I did everything right, but when I import in Metamask, all I get is an empty image. [Here](https://sepolia.etherscan.io/address/0xb9D53d659Cf041D7bA7662E0607BE5AAd373Bb32#code)'s the contract of the BasicNft I deployed during lesson 11, with which I believe there's nothing wrong. However, as you can see, I get no image:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/33223441/99291dc9-d9d1-4214-a2ee-cc76ad76d1cf)
The same thing happens to me when I do the deployment on Anvil later in the lesson:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/33223441/f2e2dca9-f23a-453b-8e7e-48fd627537f5)

I've checked the token URI for both cases (IPFS and SVG-base64) and have no clue what could be wrong...
Any ideas? Thanks beforehand!

**Reply by DukeLurkin** [reply]
Hmmm, the only other thing I did before endabling opensea and autodetect NFT was "clear activity & nonce data".  If you haven't already done that give it a try.  Nothing was wrong with my code and there's likely nothing wrong with yours either, just Metamask.

**Reply by Irwino820** [reply]
@theirrationalone Could your please please help me? I still in the same lesson with the same problem

Here is my repo: https://github.com/Irwino820/foundry-nft-f23
This is my public address: 0x47287e81090D287123E424d35DaABD2dCd933122
Look the hundred of times i try this...

**Reply by theirrationalone** [reply]
Hey @Irwino820 !, You're doing to all right. You can move on to next Lesson if you're also done with moodNFT...
Metamask team perhaps updating or doing some modifications... Now i'm also facing the same issue even for mine, @DukeLurkin ! and also your's nft. Nothing is appearing instead of NFT Name and Symbol.

Don't take it seriously, You're already done with. You can just move on to next.

;`)

**Reply by Irwino820** [reply]
Thank you so MUCH @theirrationalone 
I want to be sure that i wasn't making any mistake o error in my code, the svg NFT was successfully deployed and shown in my MM with Sepolia and Anvil.

Happy coding  :)))

**Reply by Ynyesto** [reply]
@DukeLurkin I did that and it still didn't work, so I think I'll just move on to lesson 12 :-( @Irwino820, did you do anything other than what has been said in this thread??

**Reply by theirrationalone** [reply]
Heyyy @arynyestos ! Have u tried pinning your own json file as @DukeLurkin ! been trying thus faaaaar???

**Reply by DukeLurkin** [reply]
@arynyestos the only other thing I did, and many times in many ways, was alter the PUG= "https://ipfs.io/..." .  In my searching, some people had success with their picture after removing the the ? and anything after it so it looks like something like this: "https://ipfs.io/ipfs/Qmc5fD7GQPtC48aBHmAwCa8L9R52jq442TxBNEcFj5WL3f" I removed the ?filename=pug.png at the end of the string.  Some people remove the https:// and found that works.  Hope that helps!

**Reply by Ynyesto** [reply]
Huh, okay, so I suppose my only chance is to redeploy the contract trying those different combinations and hope one of them works, right?

**Reply by DukeLurkin** [reply]
Yup.  Change what PUG = then run make mint again.  And just remember when you mint, the token ID number will go up when you import in Metamask every time.
